src/clj_puppetdb/vcr.clj

```diff
@@ -0,0 +1,28 @@
+(ns clj-puppetdb.vcr
+  (:require [clojure.edn :as edn]
+            [me.raynes.fs :as fs]
+            [puppetlabs.kitchensink.core :refer [utf8-string->sha1]]
+            [puppetlabs.http.client.sync :as http]
+            [puppetlabs.kitchensink.core :refer [dissoc-in]]))
+
+(defn- vcr-filename
+  [vcr-dir path]
+  (str vcr-dir "/" (utf8-string->sha1 path) ".clj"))
```

Comment:
I'm afraid this is not unique enough - imagine processing the equivalents of the following two queries:
```
curl -s --get http://localhost:8080/v3/aggregate-event-counts --data-urlencode 'query=[">", "report-receive-time", "2015-01-01T14:53:07Z"]' --data-urlencode 'summarize-by=resource'
curl -s --get http://localhost:8080/v3/aggregate-event-counts --data-urlencode 'query=[">", "report-receive-time", "2015-01-01T14:53:07Z"]' --data-urlencode 'summarize-by=certname'
```
The first one would be recorded but the second would be served from the recording of the first one - because it only differs in query parameters (precisely in the value of the ```summarize-by``` parameter) so the hash computed only from the request path is the same for both queries and so the name of the corresponding vcr files are the same.
Clearly the query parameters must be somehow included in the hash computation.
This presents some challenges though as the order of the query parameters is arbitrary, so in order to obtain stable hash values they would have to be canonicalized (sorted) in some way.
This sorting could be quite costly though as it would have to be applied to all nested structures within the parameters.
I approached the problem from another angle:
I store the entire request path and query parameters in the VCR file along with the response. (I only make sure the file name is unique - so that I don't overwrite files representing response to another query.) Then when I'm searching for a response matching current query I scan each file in the vcr-dir and check if its path and query params match those of the current query. (Note that I scan all the files in vcr-dir only once and remember what I found so that I avoid reading all the files during each request.)
You can find the implementation here: https://github.com/mruzicka/pe-state-inspector-middleware/blob/ebf1e877ed281a945041b0ac3005b3063ea60bb4/test/puppetlabs/state_inspector/middleware/test_utils.clj